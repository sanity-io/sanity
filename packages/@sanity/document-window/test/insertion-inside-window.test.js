const DocumentWindow = require('../src/DocumentWindow')
const {getSnapshotFixture} = require('./fixtures')
const {getMockClient, gatherWindows, mockMutation} = require('./helpers')

const Query = DocumentWindow.Query
let docWindow = null

afterEach(() => {
  if (docWindow) {
    docWindow.removeAllListeners()
  }
})

describe('insertion inside of range', () => {
  test('injects into window when item is added to middle of range, fills post', async () => {
    const newDocument = {
      _id: 'alpha',
      name: 'That new product',
      _updatedAt: '2017-06-02T12:00:00Z',
      __injected: true
    }

    const query = new Query().from(0).to(5).orderBy('_updatedAt', 'desc')
    const client = getMockClient({
      responses: [getSnapshotFixture(0, 10)],
      events: [mockMutation(newDocument)]
    })

    docWindow = new DocumentWindow({client, query})
    const windows = await gatherWindows(docWindow, 2)

    expect(windows[0]).toMatchSnapshot('01. a) window prior to inject')
    expect(windows[1]).toMatchSnapshot('01. b) window after injection')
    expect(docWindow).toMatchSnapshot('01. c) buffers after injection')
  })

  test('injects into window when item is added to beginning of range, fills post', async () => {
    const newDocument = {
      _id: 'abacus',
      name: 'That new product',
      _updatedAt: '2017-06-02T12:00:00Z',
      __injected: true
    }

    const query = new Query().from(0).to(5).orderBy('_updatedAt', 'desc')
    const client = getMockClient({
      responses: [getSnapshotFixture(0, 10)],
      events: [mockMutation(newDocument)]
    })

    docWindow = new DocumentWindow({client, query})
    const windows = await gatherWindows(docWindow, 2)

    expect(windows[0]).toMatchSnapshot('02. a) window prior to inject')
    expect(windows[1]).toMatchSnapshot('02. b) window after injection')
    expect(docWindow).toMatchSnapshot('02. c) buffers after injection')
  })

  test('injects into window when window only contains a single item (end)', async () => {
    const newDocument = {
      _id: 'alpha',
      name: 'That new product',
      _updatedAt: '2017-06-02T12:00:00Z',
      __injected: true
    }

    const query = new Query().from(0).to(5).orderBy('_updatedAt', 'desc')
    const client = getMockClient({
      responses: [getSnapshotFixture(0, 1)],
      events: [mockMutation(newDocument)]
    })

    docWindow = new DocumentWindow({client, query})
    const windows = await gatherWindows(docWindow, 2)

    expect(windows[0]).toMatchSnapshot('03. a) window prior to inject')
    expect(windows[1]).toMatchSnapshot('03. b) window after injection')
    expect(docWindow).toMatchSnapshot('03. c) buffers after injection')
  })

  test('injects into window when window only contains a single item (start)', async () => {
    const newDocument = {
      _id: 'abacus',
      name: 'That new product',
      _updatedAt: '2017-06-02T12:00:00Z',
      __injected: true
    }

    const query = new Query().from(0).to(5).orderBy('_updatedAt', 'desc')
    const client = getMockClient({
      responses: [getSnapshotFixture(0, 1)],
      events: [mockMutation(newDocument)]
    })

    docWindow = new DocumentWindow({client, query})
    const windows = await gatherWindows(docWindow, 2)

    expect(windows[0]).toMatchSnapshot('04. a) window prior to inject')
    expect(windows[1]).toMatchSnapshot('04. b) window after injection')
    expect(docWindow).toMatchSnapshot('04. c) buffers after injection')
  })

  test('injects into window when window is empty', async () => {
    const newDocument = {
      _id: 'alpha',
      name: 'That new product',
      _updatedAt: '2017-06-02T12:00:00Z',
      __injected: true
    }

    const query = new Query().from(0).to(5).orderBy('_updatedAt', 'desc')
    const client = getMockClient({
      responses: [[]],
      events: [mockMutation(newDocument)]
    })

    docWindow = new DocumentWindow({client, query})
    const windows = await gatherWindows(docWindow, 2)

    expect(windows[0]).toMatchSnapshot('05. a) window prior to inject')
    expect(windows[1]).toMatchSnapshot('05. b) window after injection')
    expect(docWindow).toMatchSnapshot('05. c) buffers after injection')
  })
})
