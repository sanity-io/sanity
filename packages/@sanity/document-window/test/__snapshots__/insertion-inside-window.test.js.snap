// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`01. a) window prior to inject 1`] = `
Array [
  Object {
    "_id": "abcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 0",
    "productIndex": 0,
  },
  Object {
    "_id": "bcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 1",
    "productIndex": 1,
  },
  Object {
    "_id": "cdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 2",
    "productIndex": 2,
  },
  Object {
    "_id": "defghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 3",
    "productIndex": 3,
  },
  Object {
    "_id": "efghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 4",
    "productIndex": 4,
  },
]
`;

exports[`01. b) window after injection 1`] = `
Array [
  Object {
    "_id": "abcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 0",
    "productIndex": 0,
  },
  
  ----[INJECTED]----
  Object {
    "_id": "alpha",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "That new product",
  }
  ---[/INJECTED]----

,
  Object {
    "_id": "bcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 1",
    "productIndex": 1,
  },
  Object {
    "_id": "cdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 2",
    "productIndex": 2,
  },
  Object {
    "_id": "defghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 3",
    "productIndex": 3,
  },
]
`;

exports[`01. c) buffers after injection 1`] = `
[=== Pre ===]
Array []
[=== /Pre ===]

[=== Window ===]
Array [
  Object {
    "_id": "abcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 0",
    "productIndex": 0,
  },
  
  ----[INJECTED]----
  Object {
    "_id": "alpha",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "That new product",
  }
  ---[/INJECTED]----

,
  Object {
    "_id": "bcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 1",
    "productIndex": 1,
  },
  Object {
    "_id": "cdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 2",
    "productIndex": 2,
  },
  Object {
    "_id": "defghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 3",
    "productIndex": 3,
  },
]
[=== /Window ===]

[=== Post ===]
Array [
  Object {
    "_id": "efghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 4",
    "productIndex": 4,
  },
  Object {
    "_id": "fghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 5",
    "productIndex": 5,
  },
  Object {
    "_id": "ghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 6",
    "productIndex": 6,
  },
  Object {
    "_id": "hijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 7",
    "productIndex": 7,
  },
  Object {
    "_id": "ijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 8",
    "productIndex": 8,
  },
  Object {
    "_id": "jklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 9",
    "productIndex": 9,
  },
]
[=== /Post ===]

`;

exports[`02. a) window prior to inject 1`] = `
Array [
  Object {
    "_id": "abcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 0",
    "productIndex": 0,
  },
  Object {
    "_id": "bcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 1",
    "productIndex": 1,
  },
  Object {
    "_id": "cdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 2",
    "productIndex": 2,
  },
  Object {
    "_id": "defghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 3",
    "productIndex": 3,
  },
  Object {
    "_id": "efghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 4",
    "productIndex": 4,
  },
]
`;

exports[`02. b) window after injection 1`] = `
Array [
  
  ----[INJECTED]----
  Object {
    "_id": "abacus",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "That new product",
  }
  ---[/INJECTED]----

,
  Object {
    "_id": "abcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 0",
    "productIndex": 0,
  },
  Object {
    "_id": "bcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 1",
    "productIndex": 1,
  },
  Object {
    "_id": "cdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 2",
    "productIndex": 2,
  },
  Object {
    "_id": "defghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 3",
    "productIndex": 3,
  },
]
`;

exports[`02. c) buffers after injection 1`] = `
[=== Pre ===]
Array []
[=== /Pre ===]

[=== Window ===]
Array [
  
  ----[INJECTED]----
  Object {
    "_id": "abacus",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "That new product",
  }
  ---[/INJECTED]----

,
  Object {
    "_id": "abcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 0",
    "productIndex": 0,
  },
  Object {
    "_id": "bcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 1",
    "productIndex": 1,
  },
  Object {
    "_id": "cdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 2",
    "productIndex": 2,
  },
  Object {
    "_id": "defghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 3",
    "productIndex": 3,
  },
]
[=== /Window ===]

[=== Post ===]
Array [
  Object {
    "_id": "efghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 4",
    "productIndex": 4,
  },
  Object {
    "_id": "fghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 5",
    "productIndex": 5,
  },
  Object {
    "_id": "ghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 6",
    "productIndex": 6,
  },
  Object {
    "_id": "hijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 7",
    "productIndex": 7,
  },
  Object {
    "_id": "ijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 8",
    "productIndex": 8,
  },
  Object {
    "_id": "jklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 9",
    "productIndex": 9,
  },
]
[=== /Post ===]

`;

exports[`03. a) window prior to inject 1`] = `
Array [
  Object {
    "_id": "abcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 0",
    "productIndex": 0,
  },
]
`;

exports[`03. b) window after injection 1`] = `
Array [
  Object {
    "_id": "abcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 0",
    "productIndex": 0,
  },
  
  ----[INJECTED]----
  Object {
    "_id": "alpha",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "That new product",
  }
  ---[/INJECTED]----

,
]
`;

exports[`03. c) buffers after injection 1`] = `
[=== Pre ===]
Array []
[=== /Pre ===]

[=== Window ===]
Array [
  Object {
    "_id": "abcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 0",
    "productIndex": 0,
  },
  
  ----[INJECTED]----
  Object {
    "_id": "alpha",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "That new product",
  }
  ---[/INJECTED]----

,
]
[=== /Window ===]

[=== Post ===]
Array []
[=== /Post ===]

`;

exports[`04. a) window prior to inject 1`] = `
Array [
  Object {
    "_id": "abcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 0",
    "productIndex": 0,
  },
]
`;

exports[`04. b) window after injection 1`] = `
Array [
  
  ----[INJECTED]----
  Object {
    "_id": "abacus",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "That new product",
  }
  ---[/INJECTED]----

,
  Object {
    "_id": "abcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 0",
    "productIndex": 0,
  },
]
`;

exports[`04. c) buffers after injection 1`] = `
[=== Pre ===]
Array []
[=== /Pre ===]

[=== Window ===]
Array [
  
  ----[INJECTED]----
  Object {
    "_id": "abacus",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "That new product",
  }
  ---[/INJECTED]----

,
  Object {
    "_id": "abcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 0",
    "productIndex": 0,
  },
]
[=== /Window ===]

[=== Post ===]
Array []
[=== /Post ===]

`;

exports[`05. a) window prior to inject 1`] = `Array []`;

exports[`05. b) window after injection 1`] = `
Array [
  
  ----[INJECTED]----
  Object {
    "_id": "alpha",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "That new product",
  }
  ---[/INJECTED]----

,
]
`;

exports[`05. c) buffers after injection 1`] = `
[=== Pre ===]
Array []
[=== /Pre ===]

[=== Window ===]
Array [
  
  ----[INJECTED]----
  Object {
    "_id": "alpha",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "That new product",
  }
  ---[/INJECTED]----

,
]
[=== /Window ===]

[=== Post ===]
Array []
[=== /Post ===]

`;
